File: mobile/src/main/java/com/bclymer/dailybudget/fragments/BudgetTransactionsFragment.java
package com.bclymer.dailybudget.fragments;

import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AbsListView;
import android.widget.BaseAdapter;
import android.widget.EditText;

import com.bclymer.dailybudget.R;
import com.bclymer.dailybudget.events.BudgetUpdatedEvent;
import com.bclymer.dailybudget.models.Budget;
import com.bclymer.dailybudget.models.Transaction;
import com.bclymer.dailybudget.utilities.ThreadManager;
import com.bclymer.dailybudget.views.TransactionView;

import java.util.ArrayList;
import java.util.List;

import butterknife.Bind;
import butterknife.OnClick;

/**
 * Created by bclymer on 9/28/2014.
 */
public class BudgetTransactionsFragment extends BaseDialogFragment {

    public static final String TAG = "BudgetTransactionsFragment";

    private static final String EXTRA_BUDGET_ID = "extra_budget_id";

    @Bind(android.R.id.list)
    protected AbsListView mListView;
    @Bind(android.R.id.empty)
    protected ViewGroup mEmptyView;
    @Bind(R.id.fragment_budget_transactions_edittext_filter)
    protected EditText mEditTextFilter;

    private TransactionAdapter mAdapter;
    private List<Transaction> mTransactionList;
    private List<Transaction> mTransactionListFiltered;

    private int mBudgetId;

    public BudgetTransactionsFragment() {
    }

    public static BudgetTransactionsFragment newInstance(int budgetId) {
        BudgetTransactionsFragment fragment = new BudgetTransactionsFragment();
        Bundle bundle = new Bundle(1);
        bundle.putInt(EXTRA_BUDGET_ID, budgetId);
        fragment.setArguments(bundle);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setMLayoutId(R.layout.fragment_budget_transactions);
        mBudgetId = getArguments().getInt(EXTRA_BUDGET_ID);
    }

    @Override
    public void onViewCreated(View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        Budget budget = Budget.getDao().queryForId(mBudgetId);

        getDialog().setTitle(budget.name);
        mTransactionList = budget.getSortedTransactions();
        mTransactionListFiltered = new ArrayList<>(mTransactionList);
        mAdapter = new TransactionAdapter();
        mListView.setAdapter(mAdapter);
        mListView.setEmptyView(mEmptyView);

        mEditTextFilter.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
            }

            @Override
            public void afterTextChanged(Editable s) {
                filterFullList();
                mAdapter.notifyDataSetChanged();
            }
        });

        getMEventBus().register(this);
    }

    @Override
    public void onDestroyView() {
        getMEventBus().unregister(this);
        super.onDestroyView();
    }

    @OnClick(R.id.fragment_budget_transactions_button_new_transaction)
    protected void clickedAddTransaction() {
        EditTransactionFragment.newInstance(mBudgetId).show(getFragmentManager(), EditTransactionFragment.TAG);
    }

    @OnClick(R.id.fragment_budget_transactions_button_stats)
    protected void clickedBudgetStats() {
        BudgetStatsFragment.Companion.newInstance(mBudgetId).show(getFragmentManager(), BudgetStatsFragment.Companion.getTAG());
    }

    private void filterFullList() {
        final String filter = mEditTextFilter.getText().toString();
        if (filter.length() == 0) {
            mTransactionListFiltered = new ArrayList<>(mTransactionList);
        }
        mTransactionListFiltered.clear();
        for (Transaction t : mTransactionList) {
            if (t.location == null) continue;
            if (t.location.toLowerCase().contains(filter.toLowerCase())) {
                mTransactionListFiltered.add(t);
            }
        }
    }

    public void onEvent(final BudgetUpdatedEvent event) {
        if (event.getBudget().id != mBudgetId) return;

        ThreadManager.runInBackgroundThenUi(new Runnable() {
            @Override
            public void run() {
                mTransactionList = event.getBudget().getSortedTransactions();
                filterFullList();
            }
        }, new Runnable() {
            @Override
            public void run() {
                if (!isVisible() || mListView == null) return;

                mAdapter.notifyDataSetChanged();
            }
        });
    }

    private class TransactionAdapter extends BaseAdapter {

        private final LayoutInflater mInflater;

        public TransactionAdapter() {
            mInflater = getActivity().getLayoutInflater();
        }

        @Override
        public int getCount() {
            return mTransactionListFiltered != null ? mTransactionListFiltered.size() : 0;
        }

        @Override
        public Transaction getItem(int position) {
            return mTransactionListFiltered.get(position);
        }

        @Override
        public long getItemId(int position) {
            return position;
        }

        @Override
        public View getView(int position, View convertView, ViewGroup parent) {
            final Transaction transaction = mTransactionListFiltered.get(position);
            final TransactionView transactionView = TransactionView.createTransactionView(mInflater, (TransactionView) convertView, parent, transaction);
            transactionView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    EditTransactionFragment.newInstance(mBudgetId, transaction.id).show(getFragmentManager(), EditTransactionFragment.TAG);
                }
            });
            return transactionView;
        }
    }
}

File: mobile/src/main/java/com/bclymer/dailybudget/fragments/BudgetTransactionsFragment.kt
package com.bclymer.dailybudget.fragments

import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.AbsListView
import android.widget.BaseAdapter
import android.widget.Button
import android.widget.EditText
import com.bclymer.dailybudget.MainActivity
import com.bclymer.dailybudget.R
import com.bclymer.dailybudget.database.BudgetRepository
import com.bclymer.dailybudget.database.TransactionRepository
import com.bclymer.dailybudget.models.Transaction
import com.bclymer.dailybudget.views.TransactionView
import com.travefy.travefy.core.bindView
import java.util.*

/**
 * Created by bclymer on 9/28/2014.
 */
class BudgetTransactionsFragment() : BaseFragment(R.layout.fragment_budget_transactions) {

    private val mListView: AbsListView by bindView(R.id.fragment_budget_transactions_listview)
    private val mEmptyView: ViewGroup by bindView(R.id.include_no_transactions_empty)
    private val mEditTextFilter: EditText by bindView(R.id.fragment_budget_transactions_edittext_filter)
    private val mButtonNewTransaction: Button by bindView(R.id.fragment_budget_transactions_button_new_transaction)
    private val mButtonStats: Button by bindView(R.id.fragment_budget_transactions_button_stats)

    private val mAdapter = TransactionAdapter()

    private var mTransactionList: List<Transaction> = listOf()
    private var mTransactionListFiltered: MutableList<Transaction> = mutableListOf()

    private var mBudgetId: Int = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        mBudgetId = arguments.getInt(EXTRA_BUDGET_ID)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        (activity as MainActivity).actionBar.title = BudgetRepository.getById(mBudgetId)?.name

        TransactionRepository.monitorTransactions(mBudgetId).subscribeOnLifecycle(onNext = {
            mTransactionList = it
            mTransactionListFiltered = ArrayList(mTransactionList)
        })

        mListView.adapter = mAdapter
        mListView.emptyView = mEmptyView

        mEditTextFilter.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {
            }

            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
            }

            override fun afterTextChanged(s: Editable) {
                filterFullList()
                mAdapter.notifyDataSetChanged()
            }
        })

        mButtonNewTransaction.setOnClickListener {
            fragmentManager.beginTransaction()
                    .add(R.id.main_activity_fragment_main, EditTransactionFragment.newInstance(mBudgetId), EditTransactionFragment.TAG)
                    .commit()
        }

        mButtonStats.setOnClickListener {
            fragmentManager.beginTransaction()
                    .add(R.id.main_activity_fragment_main, BudgetStatsFragment.newInstance(mBudgetId), BudgetStatsFragment.TAG)
                    .commit()
        }
    }

    private fun filterFullList() {
        val filter = mEditTextFilter.text.toString()
        if (filter.length == 0) {
            mTransactionListFiltered = ArrayList(mTransactionList)
        }
        mTransactionListFiltered.clear()
        for (t in mTransactionList) {
            if (t.location == null) continue
            if (t.location.toLowerCase().contains(filter.toLowerCase())) {
                mTransactionListFiltered.add(t)
            }
        }
    }

    private inner class TransactionAdapter : BaseAdapter() {

        private val mInflater: LayoutInflater by lazy {
            activity.layoutInflater
        }

        override fun getCount(): Int {
            return mTransactionListFiltered.size
        }

        override fun getItem(position: Int): Transaction {
            return mTransactionListFiltered[position]
        }

        override fun getItemId(position: Int): Long {
            return position.toLong()
        }

        override fun getView(position: Int, convertView: View?, parent: ViewGroup?): View {
            val transaction = mTransactionListFiltered[position]
            val transactionView = TransactionView.createTransactionView(mInflater, convertView as? TransactionView, parent, transaction)
            transactionView.setOnClickListener {
                fragmentManager.beginTransaction()
                        .add(R.id.main_activity_fragment_main, EditTransactionFragment.newInstance(mBudgetId, transaction.id), EditTransactionFragment.TAG)
                        .commit()
            }
            return transactionView
        }
    }

    companion object {

        val TAG = "BudgetTransactionsFragment"

        private val EXTRA_BUDGET_ID = "extra_budget_id"

        fun newInstance(budgetId: Int): BudgetTransactionsFragment {
            val fragment = BudgetTransactionsFragment()
            val bundle = Bundle(1)
            bundle.putInt(EXTRA_BUDGET_ID, budgetId)
            fragment.arguments = bundle
            return fragment
        }
    }
}
